File: LeetCode/src/string/ZigzagConversion.java
package string;

public class ZigzagConversion {

    public String convert(String s, int numRows) {
        if (numRows == 1) {
            return s;
        }

        StringBuilder[] rows = new StringBuilder[numRows];
        for (int i = 0; i < numRows; i++) {
            rows[i] = new StringBuilder();
        }

        int row = 0;
        boolean down = false;

        for (char ch : s.toCharArray()) {
            rows[row].append(ch);
            if (row == 0 || row == numRows - 1) {
                down = !down;
            }
            row += down ? 1 : -1;
        }

        StringBuilder result = new StringBuilder();
        for (StringBuilder sb : rows) {
            result.append(sb);
        }
        return result.toString();
    }

    public static void main(String[] args) {
        ZigzagConversion solution = new ZigzagConversion();
        String result1 = solution.convert("PAYPALISHIRING", 3);
        System.out.println(result1);  // Output: "PAHNAPLSIIGYIR"

        String result2 = solution.convert("PAYPALISHIRING", 4);
        System.out.println(result2);  // Output: "PINALSIGYAHRPI"
    }
}

File: LeetCode_Python/string/ZigzagConversion.py
class ZigzagConversion:
    def convert(self, s: str, numRows: int) -> str:
        if numRows == 1 or numRows >= len(s):
            return s

        rows = [""] * numRows
        cur = 0
        down = False

        for char in s:
            rows[cur] += char

            if cur == 0 or cur == numRows - 1:
                down = not down

            if down:
                cur += 1
            else:
                cur -= 1

        return "".join(rows)


# Time Complexity: O(n), where n is the length of the input string s.
# Space Complexity: O(n), as we are using a list of size n to store the rows.

if __name__ == "__main__":
    solution = ZigzagConversion()

    print("Expected Output: PAHNAPLSIIGYIR")
    print("Actual Output:", solution.convert("PAYPALISHIRING", 3))

    print("Expected Output: PINALSIGYAHRPI")
    print("Actual Output:", solution.convert("PAYPALISHIRING", 4))
